airbyte-webapp/src/components/connection/CatalogTree/CatalogSection.tsx

```diff
@@ -112,44 +112,44 @@ const CatalogSectionInner: React.FC<CatalogSectionInnerProps> = ({
       if (!config) {
         return;
       }
-
       const updatedConfig = updatePrimaryKey(config, newPrimaryKey, numberOfFieldsInStream);
-
       updateStreamWithConfig(updatedConfig);
     },
     [config, updateStreamWithConfig, numberOfFieldsInStream]
   );
 
-  const onToggleFieldSelected = (fieldPath: string[], isSelected: boolean) => {
-    const previouslySelectedFields = config?.selectedFields || [];
+  const toggleAllFieldsSelected = () => {
+    const wasFieldSelectionEnabled = config?.fieldSelectionEnabled;
+    const fieldSelectionEnabled = !wasFieldSelectionEnabled;
+    const selectedFields: string[][] = [];
 
-    if (!config?.fieldSelectionEnabled && !isSelected) {
-      // All fields in a stream are implicitly selected. When deselecting the first one, we also need to explicitly select the rest.
-      const allOtherFields = fields.filter((field: SyncSchemaField) => !isEqual(field.path, fieldPath)) ?? [];
-      const selectedFields: SelectedFieldInfo[] = allOtherFields.map((field) => ({ fieldPath: field.path }));
-      updateStreamWithConfig({
-        selectedFields,
-        fieldSelectionEnabled: true,
-      });
-    } else if (isSelected && previouslySelectedFields.length === numberOfFieldsInStream - 1) {
-      // In this case we are selecting the only unselected field
-      updateStreamWithConfig({
-        selectedFields: [],
-        fieldSelectionEnabled: false,
-      });
-    } else if (isSelected) {
-      updateStreamWithConfig({
-        selectedFields: [...previouslySelectedFields, { fieldPath }],
-        fieldSelectionEnabled: true,
-      });
-    } else {
-      updateStreamWithConfig({
-        selectedFields: previouslySelectedFields.filter((f) => !isEqual(f.fieldPath, fieldPath)) || [],
-        fieldSelectionEnabled: true,
-      });
+    // When deselecting all fields, we need to be careful not to deselect any primary keys or the cursor field
+    if (!wasFieldSelectionEnabled) {
+      if (config?.primaryKey) {
+        selectedFields.push(...config.primaryKey);
+      }
+      if (config?.cursorField) {
+        selectedFields.push(config.cursorField);
+      }
     }
+
+    updateStreamWithConfig({
+      fieldSelectionEnabled,
+      selectedFields: selectedFields.map((fieldPath) => ({ fieldPath })),
+    });
   };
 
+  const onToggleFieldSelected = useCallback(
+    (fieldPath: string[], isSelected: boolean) => {
+      if (!config) {
+        return;
+      }
+      const updatedConfig = updateFieldSelected({ config, fields, fieldPath, isSelected, numberOfFieldsInStream });
+      updateStreamWithConfig(updatedConfig);
+    },
+    [config, fields, numberOfFieldsInStream, updateStreamWithConfig]
+  );
```

Comment:
This has been extracted to `streamConfigHelpers` to make unit testing easier (see `streamConfigHelpers.test.ts`)